### ISSUE TITLE
Icons missing when using relative URL

### ISSUE THREAD

**Issue by lschr** [gitlabhq/gitlabhq#10053]
I set my relative URL to /gitlab by following the instructions in the documentation. However, now quite a lof of icons are missing in the web interface.

I guess the problem is that the URL of FontAwesome is not updated in _application-9880f992b4a3c109f2a12a3f4553d0fe.css._ There it says
`/@font-face{font-family:'FontAwesome';src:url("/assets/fontawesome-webfont-39aed32f180471d5eb26b657310bc531.eot?v=4.5.0")` and so on instead of `/gitlab/assets`.

I am using gitlab-ce from the apt repository.

**Comment by artem-sidorenko** [comment]
@lschr it looks like asset (re-)complication was not successful. Did you observe some errors in  `gitlab-ctl reconfigure` when you enabled the relative url?

You can try to recompile the assets manually by following [this steps](http://doc.gitlab.com/omnibus/settings/configuration.html#relative-url-troubleshooting)

**Comment by lschr** [comment]
I did not see any errors. I also tried the troubleshooting steps, but they didn't help.

Btw, it looks like all URLs in the CSS file are wrong, not only the one for the font.

**Comment by artem-sidorenko** [comment]
@lschr can you please crosscheck
- you use the Gitlab CE 8.5
- external_url in `gitlab.rb` is like `http(s)://example.com/gitlab`
- you executed `gitlab-ctl reconfigure`
- you have a file `/var/opt/gitlab/gitlab-rails/etc/relative_url.rb` with following content
```
# This file is managed by gitlab-ctl. Manual changes will be
# erased! To change the contents below, edit /etc/gitlab/gitlab.rb
# and run `sudo gitlab-ctl reconfigure`.

Rails.application.configure do
  config.relative_url_root = "/gitlab"
end
```

**Comment by lschr** [comment]
`dpkg -l` tells me
```ii  gitlab-ce            8.5.0-ce.0      amd64           GitLab Community Edition and GitLab CI (includ```

`grep ^external_url /etc/gitlab/gitlab.rb` gives
```external_url 'http://gitlab/gitlab'```

I executed `gitlab-ctl reconfigure` and `gitlab-ctl restart` several times

Finally, `cat /var/opt/gitlab/gitlab-rails/etc/relative_url.rb` outputs
```
# This file is managed by gitlab-ctl. Manual changes will be
# erased! To change the contents below, edit /etc/gitlab/gitlab.rb
# and run `sudo gitlab-ctl reconfigure`.

Rails.application.configure do
  config.relative_url_root = "/gitlab"
end
```

I also tried reinstalling _gitlab-ce_ to no avail.

**Comment by artem-sidorenko** [comment]
@lschr I will try to reproduce it today with `external_url 'http://gitlab/gitlab'`. Can you please tell me the Ubuntu/Debian version you have?

is it possible for you to provide a full log of `gitlab-ctl reconfigure`, but completely on the fresh installation without any sensitive information or similar things? So, what I mean:
- `gitlab-ctl stop`
- uninstall gitlab
- `rm -rf /var/opt/gitllab`
- install gitlab
- configure `external_url 'http://gitlab/gitlab'` in gitlab.rb, but without any further settings
- do `gitlab-ctl reconfigure | tee log`
- upload this log here as attachment

**Comment by lschr** [comment]
I am using an LXC container with Ubuntu 14.04, so I can even install in a fresh container. After installing, it prompts me to run `gitlab-ctl reconfigure`. Should I change the ``external_url`` before that and post the log of `gitlab-ctl reconfigure`, or should I run it once, change the URL, run it again and post the log of that?

**Comment by artem-sidorenko** [comment]
@lschr after a fresh installation, can you do both?:)

**Comment by artem-sidorenko** [comment]
@Ischr I guess I could reproduce that on both centos and ubuntu:

```bash
$ grep font /var/log/gitlab/nginx/gitlab_access.log 
10.0.2.2 - - [25/Feb/2016:17:51:45 +0000] "GET /assets/fontawesome-webfont-be0584438c9b22a95d72e4aaebfe12ed.woff2?v=4.5.0 HTTP/1.1" 200 66624 "http://localhost:2082/gitlab/assets/application-9880f992b4a3c109f2a12a3f4553d0fe.css" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:44.0) Gecko/20100101 Firefox/44.0"
10.0.2.2 - - [25/Feb/2016:18:00:27 +0000] "GET /assets/fontawesome-webfont-be0584438c9b22a95d72e4aaebfe12ed.woff2?v=4.5.0 HTTP/1.1" 200 66674 "http://localhost:2082/gitlab/assets/application-9880f992b4a3c109f2a12a3f4553d0fe.css" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/48.0.2564.116 Chrome/48.0.2564.116 Safari/537.36"
$ grep "^external_url" /etc/gitlab/gitlab.rb 
external_url 'http://gitlab/gitlab'
```

First problem:
- In my source based installation this css file has the `/gitlab/assets` for this font but the omnibus installation in 8.5.0 has `/assets` after asset recompile.

Second problem which might make it difficult to reproduce this thing:
- nginx in omnibus serves the `/assets`. But in your setup you probably have a reverse proxy which passes the request to nginx running in your lxc container. This reverse proxy doesn't pass the `/assets` and expects it to be `/gitlab/assets` (Which would be my target setup too)

I'll have a deeper look in the next days

**Comment by artem-sidorenko** [comment]
- we use `initializers/relative_url.rb` for configuration of relative URL in omnibus
- initializers are not loaded during the rake task `assets:precompile`, so assets generation runs without relative url configuration in Rails
- `config.assets.initialize_on_precompile` [does not exist anymore](https://github.com/rails/rails/commit/2d5a6de4e227ee97ccbea691c69728b7e31bf678)
- in source based installations this configuration is done in `application.rb`, which gets loaded in the Rakefile, so it works fine there

Due to the nginx thing described above + browser caching we totally missed this problem during tests and review

I see multiple possible ways to resolve this:
 - to add `require File.expand_path('../initializers/relative_url', __FILE__)` with a check if this file exists to `application.rb`
 - to add `require File.expand_path('../config/initializers/relative_url', __FILE__)`  with a check if this file exists to `Rakefile`

@stanhu @maxlazio what do you think?

**Comment by artem-sidorenko** [comment]
https://gitlab.com/gitlab-org/omnibus-gitlab/merge_requests/662
https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/2979

**Comment by lschr** [comment]
@artem-sidorenko It seems you were a lot quicker than me (`gitlab-ctl` seems to suppress all output when it is piped, so `gitlab-ctl reconfigure | tee log` didn't work as expected). You are totally right about me using a reverse proxy.

Thank you very much for taking care of the problem so quickly!

**Comment by artem-sidorenko** [comment]
@lschr as I contributed this I have also to take care about the consequences and not simply leave upstream alone here:) So your thank you should be directed to @stanhu, without his ping I would not be aware of this issue

And this thing is really good for installation I maintain too, as I would definitely catch this problem during the migration today and it would be a mess to troubleshoot it in our complex setup.

**Comment by rymai** [comment]
@lschr This should be fixed in latest 8.5.2 release.
